src/storage-operators/src/s3_oneshot_sink/parquet.rs

```diff
@@ -7,28 +7,337 @@
+/// for a single parquet file.
+struct ParquetFile {
+    writer: ArrowWriter<Vec<u8>>,
+    uploader: S3MultiPartUploader,
+    max_file_size: u64,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ParquetFileError {
+    #[error("RecordBatch exceeds max file size limit")]
+    RecordBatchExceedsMaxFileLimit,
+    #[error(transparent)]
+    UploaderError(#[from] S3MultiPartUploadError),
+    #[error(transparent)]
+    Other(#[from] anyhow::Error),
+}
+
+impl ParquetFile {
+    async fn new(
+        schema: Schema,
+        bucket: String,
+        key: String,
+        sdk_config: Arc<SdkConfig>,
+        max_file_size: u64,
+    ) -> Result<Self, anyhow::Error> {
+        // TODO: Build a WriterProperties struct with the desired parquet file properties
+        // and the correct buffering strategy (configuring the row group size, etc.).
+        let writer = ArrowWriter::try_new(Vec::new(), schema.into(), None)?;
+        let uploader = S3MultiPartUploader::try_new(
+            sdk_config.as_ref(),
+            bucket,
+            key,
+            S3MultiPartUploaderConfig {
+                part_size_limit: UPLOAD_MULTIPART_PART_SIZE_LIMIT.as_u64(),
+                // We are already enforcing the max size ourselves since we can't handle a
+                // `UploadExceedsMaxFileLimit` error from the library without ruining our
+                // arrow writer state, so we set the max size enforced by the uploader to the max
+                // file size it will allow based on the part size limit.
+                // This is known to be greater than the `MAX_S3_SINK_FILE_SIZE` enforced during
+                // sink creation.
+                file_size_limit: UPLOAD_MULTIPART_PART_SIZE_LIMIT
+                    .as_u64()
+                    .checked_mul(AWS_S3_MAX_PART_COUNT.try_into().expect("known safe"))
+                    .expect("known safe"),
+            },
+        )
+        .await?;
+        Ok(Self {
+            writer,
+            uploader,
+            max_file_size,
+        })
+    }
+
+    async fn finish(mut self) -> Result<CompletedUpload, anyhow::Error> {
+        let buffer = self.writer.into_inner()?;
+        self.uploader.buffer_chunk(buffer.as_slice()).await?;
+        Ok(self.uploader.finish().await?)
+    }
+
+    /// Writes an arrow Record Batch to the parquet writer, then flushes the writer's buffer to
+    /// the uploader which may trigger an upload.
+    async fn write_arrow_batch(
+        &mut self,
+        record_batch: &RecordBatch,
+    ) -> Result<(), ParquetFileError> {
+        // Check if the current batch would exceed the file size limit
+        // This is an estimate since RecordBatch.get_array_memory_size() doesn't equal the
+        // actual size of the serialized data.
+        if u64::cast_from(
+            self.writer.bytes_written()
+                + self.writer.in_progress_size()
+                + record_batch.get_array_memory_size(),
```

Comment:
I think in general this specific check is fundamental to the correct operation of the sink, so each component should be documented! I think this is related to the above question I had about handling this error, which comes down: are we overcounting or undercounting, and how does that relate to the flow of the code itself?